File: src/display/intensity.js
import consts from '/src/consts.json';

const intensities = {
  'land': {
    'energy': [0, 0.001, 0.01, 0.1],
    'calories': [0, 0.001, 0.002, 0.01],
  },
  'energy': {
    'energy': [0, 0.001, 0.01, 0.1], // TODO EROI
    'calories': [0, 0.00015, 0.0005, 0.001],
  },
  'water': {
    'energy': [0, 1, 2, 5],
    'calories': [0, 1, 2, 3],
  },
  'emissions': {
    'energy': [-2000, 0, 200, 800],
    'calories': [-1, 0, 0.5, 1],
  },
  'biodiversity': {
    'energy': [0, 1, 2, 3],
    'calories': [0, 1, 2, 3],
  },
};

function intensity(val, key, type) {
  let stops;
  if (key in intensities) {
    stops = intensities[key][type];
  } else {
    stops = consts.demand_levels.map((o) => Math.floor(o[key]));
  }
  for (let i = 0; i < stops.length - 1; i++) {
    if (val >= stops[i] && val < stops[i+1]) {
      return i+1;
    }
  }
  return stops.length;
}

function scale(val, key) {
  switch (key) {
    case 'outlook': return Math.round(val/consts.base_outlook * 4);
    case 'extinction': return Math.round(val/60 * 4);
    case 'habitability': return Math.round(val/consts.base_habitability * 4);
    case 'world_outlook': return Math.round(val/(consts.base_outlook+consts.base_world_outlook) * 4);
  }
}

function color(intensity, invert) {
  intensity = invert ? 5 - intensity : intensity;
  if (intensity <= 1) {
    return '#43CC70';
  } else if (intensity === 2) {
    return '#FBC011';
  } else if (intensity === 3) {
    return '#f28435';
  } else {
    return '#EF3838';
  }
}

function describe(intensity) {
  if (intensity === 0) {
    return 'Very Low';
  } else if (intensity === 1) {
    return 'Low';
  } else if (intensity === 2) {
    return 'Moderate';
  } else if (intensity === 3) {
    return 'High';
  } else {
    return 'Very High';
  }
}

export default {
  intensity, color, scale, describe
}

File: hes-game/src/display/intensity.rs
use crate::{state, state::GameState};
use hes_engine::kinds::Output;
use leptos::*;

use super::{Impact, OutputKind};

const BASE_WORLD_OUTLOOK: f32 = 20.;
const BASE_REGIONAL_OUTLOOK: f32 = 10.;
const BASE_REGIONAL_HABITABILITY: f32 = 10.;

pub enum Variable {
    Outlook,
    Extinction,
    Habitability,
    WorldOutlook,
    Warming,
}

fn impact_stops(key: Impact, kind: OutputKind) -> [f32; 4] {
    match key {
        Impact::Land => match kind {
            OutputKind::Energy => [0., 0.001, 0.01, 0.1],
            OutputKind::Calories => [0., 0.001, 0.002, 0.01],
        },
        Impact::Energy => match kind {
            OutputKind::Energy => [0., 0.001, 0.01, 0.1], // TODO EROI
            OutputKind::Calories => [0., 0.00015, 0.0005, 0.001],
        },
        Impact::Water => match kind {
            OutputKind::Energy => [0., 1., 2., 5.],
            OutputKind::Calories => [0., 1., 2., 3.],
        },
        Impact::Emissions => match kind {
            OutputKind::Energy => [-2000., 0., 200., 800.],
            OutputKind::Calories => [-1., 0., 0.5, 1.],
        },
        Impact::Biodiversity => match kind {
            OutputKind::Energy => [0., 1e-15, 1e-14, 1.5e-14],
            OutputKind::Calories => [0., 1e-16, 1e-15, 1e-14],
        },
        Impact::Electricity | Impact::Fuel => state!(|state, ui| {
            let output = key.as_output().expect("Checked they're valid outputs");
            state.world.demand_by_income_levels(output)
        })(),
    }
}

pub fn impact_intensity(val: f32, key: Impact, kind: OutputKind) -> usize {
    let stops = impact_stops(key, kind);
    stops
        .windows(2)
        .filter(|stops| val >= stops[0] && val < stops[1])
        .count()
}

fn output_stops(key: Output) -> [f32; 4] {
    state!(|state, ui| state.world.demand_by_income_levels(key))()
}

pub fn output_intensity(val: f32, key: Output) -> usize {
    let stops = output_stops(key);
    stops
        .windows(2)
        .filter(|stops| val >= stops[0] && val < stops[1])
        .count()
}

pub const N_PIPS: usize = 5;

pub fn color(mut intensity: usize, invert: bool) -> &'static str {
    if invert {
        intensity = N_PIPS - intensity;
    }
    if intensity <= 1 {
        "#2FE863"
    } else if intensity == 2 {
        "#FBC011"
    } else if intensity == 3 {
        "#f28435"
    } else {
        "#EF3838"
    }
}

pub fn describe(intensity: usize) -> &'static str {
    match intensity {
        0 => "Very Low",
        1 => "Low",
        2 => "Moderate",
        3 => "High",
        _ => "Very High",
    }
}

pub fn scale(val: f32, key: Variable) -> usize {
    let val = match key {
        Variable::Outlook => (val / BASE_REGIONAL_OUTLOOK * 4.).round().max(1.),
        Variable::Extinction => (val / 100. * 4.).round().max(0.),
        Variable::Habitability => (val / BASE_REGIONAL_HABITABILITY * 4.).round().max(0.),
        Variable::WorldOutlook => (val / (BASE_REGIONAL_OUTLOOK + BASE_WORLD_OUTLOOK) * 4.)
            .round()
            .max(1.),
        Variable::Warming => val.floor() + 1.,
    };
    val as usize
}
